test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriParser/SelectExpandFunctionalTests.cs

```diff
@@ -1273,6 +1273,71 @@ public void SelectAndExpandShouldWorkOnSelectOpenProperty()
+            Assert.Equal(expectedRight, right);
+            Assert.Equal(expectedLeft, left);
+        }
+
+        [Fact]
+        public void DollarItinFilterInsideNestedExpandShouldReferenceResourcePathEntity()
+        {
+            // $it/ID references PersonType since the resource is a People EntitySet.
+            IEdmStructuredType expectedRight = (IEdmStructuredType)HardCodedTestModel.GetPersonType();
+
+            // $filter=ID1 references LionsISaw whose type in LionType.
+            IEdmStructuredType expectedLeft = (IEdmStructuredType)HardCodedTestModel.GetLionType();
+            SelectExpandClause clause = RunParseSelectExpand("", "MyDog($select=Color;$expand=LionsISaw($filter=ID1 eq $it/ID))", HardCodedTestModel.GetPersonType(), HardCodedTestModel.GetPeopleSet());
+
+            IEdmStructuredType right = ((((clause.SelectedItems.First() as ExpandedNavigationSelectItem).SelectAndExpand.SelectedItems.First() as ExpandedNavigationSelectItem).FilterOption.Expression as BinaryOperatorNode).Right as SingleValuePropertyAccessNode).Property.DeclaringType;
+            IEdmStructuredType left = ((((clause.SelectedItems.First() as ExpandedNavigationSelectItem).SelectAndExpand.SelectedItems.First() as ExpandedNavigationSelectItem).FilterOption.Expression as BinaryOperatorNode).Left as SingleValuePropertyAccessNode).Property.DeclaringType;
+            Assert.Equal(expectedRight, right);
+            Assert.Equal(expectedLeft, left);
+        }
+
+        [Fact]
+        public void DollarItinFilterInsideMultiNestedExpandShouldReferenceResourcePathEntity()
+        {
+            // $it/ID references PersonType since the resource is a People EntitySet
+            IEdmStructuredType expectedRight = (IEdmStructuredType)HardCodedTestModel.GetPersonType();
+
+            // $filter=ID references MyPaintings whose type in PaintingType.
+            IEdmStructuredType expectedLeft = (IEdmStructuredType)HardCodedTestModel.GetPaintingType();
+            SelectExpandClause clause = RunParseSelectExpand("", "MyDog($select=Color;$expand=MyPeople($expand=MyPaintings($filter=ID eq $it/ID)))", HardCodedTestModel.GetPersonType(), HardCodedTestModel.GetPeopleSet());
+
+            IEdmStructuredType right = (((((clause.SelectedItems.First() as ExpandedNavigationSelectItem).SelectAndExpand.SelectedItems.First() as ExpandedNavigationSelectItem).SelectAndExpand.SelectedItems.First() as ExpandedNavigationSelectItem).FilterOption.Expression as BinaryOperatorNode).Right as SingleValuePropertyAccessNode).Property.DeclaringType;
+            IEdmStructuredType left = (((((clause.SelectedItems.First() as ExpandedNavigationSelectItem).SelectAndExpand.SelectedItems.First() as ExpandedNavigationSelectItem).SelectAndExpand.SelectedItems.First() as ExpandedNavigationSelectItem).FilterOption.Expression as BinaryOperatorNode).Left as SingleValuePropertyAccessNode).Property.DeclaringType;
+            Assert.Equal(expectedRight, right);

```

Comment:
It's too long to maintain and read. Better to separate multiple lines?

Reply:
Done.
#Resolved